merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/InsertConstraintRunsAction.java

```diff
@@ -15,6 +15,8 @@
   private static final @Language("SQL") String sql = """
     insert into constraint_run (constraint_id, constraint_definition, simulation_dataset_id, results)
     values (?, ?, ?, ?::json)
+    on conflict (constraint_id, constraint_definition, simulation_dataset_id) do update
+    set results = excluded.results;
```

Comment:
Got it, I'll revert my changes to the SQL. Just to clarify, given my limited knowledge of databases, is it considered better practice to embed logic like this directly in the database table rather than attempting to achieve the same action through SQL calls?

Reply:
I will add an e2e test as well

Reply:
> Just to clarify, given my limited knowledge of databases, is it considered better practice to embed logic like this directly in the database table rather than attempting to achieve the same action through SQL calls?

If by "logic like this", you mean the choice to use an upsert vs fixing the function, it is better practice to fix the bugged function, as by not doing so you're actually redoing work that has already been done (that's why you were seeing the conflict here -- Constraints will only upload the results it didn't get from the cache)

If by "logic like this", you mean having the `definition_outdated` column, for constraints caching we decided it was clearer to have a boolean on `constraint_run` indicating if the definition it had run with was up-to-date rather than having external tools join on the `constraints` table in order to figure that out. For keeping this boolean up to date, it is much more efficient to use a trigger on the `constraints` table than trying to make SQL calls from external Java code.

Did that explain it or did I misunderstand the question?

Reply:
You did, thank you, for answering the question.

Reply:
Done, Updated the `up.sql` and removed the modified sql expression

Reply:
The `up.sql` can't be updated, as we've already released versions of Aerie with that file. Instead, it needs to be done in a new migration.

Reply:
Additionally, the actual function in the DB needs to be changed (it is in `constraint.sql`).

Reply:
For an explanation for future reference: when DB code changes, it needs to be reflected both in the actual `.sql` files (so that new DBs are initialized with the most up-to-date-state), and in a new migration (so that existing DBs can be caught up). Generally, we don't update old migrations, especially those that have already been included in a release, unless it's to fix a bug that would prevent the migration from running.

Reply:
Got it, thank you for letting me know, I wasn't even thinking about the constraint.sql :P

Reply:
Took a couple of tires but I got it working. Sorry for the misunderstanding and delay.

Reply:
The file `merlin-server/sql/merlin/applied_migrations.sql` needs the `call migrations.mark_migration_applied('34');` statement added there as well.

Reply:
Now it is fully done :)